src/client/particles.cpp

```diff
@@ -31,635 +29,808 @@ with this program; if not, write to the Free Software Foundation, Inc.,
+
+		const struct TileAnimationParams &anim = event->add_particlespawner.animation;
+		if (anim.type != TAT_NONE) {
+			scene::IParticleAffector *animation_affector =
+				new AnimationAffector(ps, anim);
+			ps->addAffector(animation_affector);
+			animation_affector->drop();
 		}
-		case CE_ADD_PARTICLESPAWNER: {
-			{
-				MutexAutoLock lock(m_spawner_list_lock);
-				if (m_particle_spawners.find(event->add_particlespawner.id) !=
-						m_particle_spawners.end()) {
-					delete m_particle_spawners.find(event->add_particlespawner.id)->second;
-					m_particle_spawners.erase(event->add_particlespawner.id);
-				}
-			}
 
-			video::ITexture *texture =
-				client->tsrc()->getTextureForMesh(*(event->add_particlespawner.texture));
-
-			auto toadd = new ParticleSpawner(client, player,
-					event->add_particlespawner.amount,
-					event->add_particlespawner.spawntime,
-					*event->add_particlespawner.minpos,
-					*event->add_particlespawner.maxpos,
-					*event->add_particlespawner.minvel,
-					*event->add_particlespawner.maxvel,
-					*event->add_particlespawner.minacc,
-					*event->add_particlespawner.maxacc,
-					event->add_particlespawner.minexptime,
-					event->add_particlespawner.maxexptime,
-					event->add_particlespawner.minsize,
-					event->add_particlespawner.maxsize,
-					event->add_particlespawner.collisiondetection,
+		if (event->add_particlespawner.collisiondetection) {
+			scene::IParticleAffector *collision_affector =
+				new CollisionAffector(m_env,
 					event->add_particlespawner.collision_removal,
 					event->add_particlespawner.object_collision,
-					event->add_particlespawner.attached_id,
-					event->add_particlespawner.vertical,
-					texture,
-					event->add_particlespawner.animation,
-					event->add_particlespawner.glow,
-					this);
-
-			/* delete allocated content of event */
-			delete event->add_particlespawner.minpos;
-			delete event->add_particlespawner.maxpos;
-			delete event->add_particlespawner.minvel;
-			delete event->add_particlespawner.maxvel;
-			delete event->add_particlespawner.minacc;
-			delete event->add_particlespawner.texture;
-			delete event->add_particlespawner.maxacc;
-
-			{
-				MutexAutoLock lock(m_spawner_list_lock);
-				m_particle_spawners[event->add_particlespawner.id] = toadd;
-			}
-			break;
+					event->add_particlespawner.bounce_fraction,
+					event->add_particlespawner.bounce_threshold);
+			ps->addAffector(collision_affector);
+			collision_affector->drop();
+		}
+
+		scene::IParticleAffector *acceleration_affector =
+			new AccelerationAffector(*event->add_particlespawner.minacc,
+				*event->add_particlespawner.maxacc);
+		ps->addAffector(acceleration_affector);
+		acceleration_affector->drop();
+
+
+		scene::IParticleAffector *lighting_affector =
+			new LightingAffector(m_env, client, 0);
+		ps->addAffector(lighting_affector);
+		lighting_affector->drop();
+
+		// Delete allocated content of event
+		delete event->add_particlespawner.minpos;
```

Comment:
I don't like this, but it was there before I guess

Reply:
For this, the ClientEvent would have to be restructured.

Reply:
our clientevent model using a union is just painful, it should use inheritance